### DISCUSSION TITLE
Help me understand how to use a custom analyzer

### DISCUSSION THREAD

**Discussion by cadilhac** [ravendb/ravendb#16629] [Q&A]
I am using a FrenchAnalyzer to search some french text. I successully built the corresponding assembly and installed it on the RavenDb server. I declared the fields that have to use this analyzer and so far so good. Indexing works. Searching is problematic though.

With the french analyzer, words are stemmed before being stored as terms. For instance, the word "créativité" is stored as "créativ" in the index. I understand by doing this, I could search some words beginning by this term. 

However, I'm wondering how I can search with a term without the diacritics. In the facts, searching "créativité" will work, but searching "creativite" won't.

I tried adding a ASCIIFoldingFilter in the analyzer but then stemming won't work well. "creativ" would be stored in the index, but searching "creativite" will return no result.

**Comment by mcrio** [comment]
Hi @cadilhac Would you mind sharing the analyzer code. I'll be doing something similar..

How do you do the stemming, is it custom code you wrote or part of a Lucene Analyzer your class extends? 

For testing purposes, you could normalize and remove the diacritics manually in your anaylzer by using the c# string Normalize method. You'd normalize to FormD so `é` is decomposed into `e` and the diacritic. Then you'd filter out all characters which are of category:
```
string str = "créativité";
var b = new StringBuilder();
foreach (var c in str.Normalize(NormalizationForm.FormD).EnumerateRunes())
{
	if (Rune.GetUnicodeCategory(c) != UnicodeCategory.NonSpacingMark)
	{
		b.Append(c);
	}
}
var normalized = b.ToString();
Console.WriteLine(normalized);   // creativite
```

**Reply by cadilhac** [reply]
Hi,

Here is the [file](https://gist.github.com/cadilhac/1ab2c9807b8ca4ffcba379b5ec4c8db6) I use.

I can't use the method you describe, because diacritics have to be used for stemming. The problem is when searching and being lazy when not typing diacritics.

**Reply by ayende** [reply]
Use ASCII folding after the stemming?

**Reply by cadilhac** [reply]
I tried but the change will only store stemmed words without the diacritics. It won't help the search part since searching for a keyword without diacritics won't be stemmed the same way.

**Reply by ayende** [reply]
When you return the term, you can return multiple terms for a single word.
So do that for with and without the diatrics.

**Reply by cadilhac** [reply]
What do you mean by "When you return the term"?

If that means that from the word "créativité", both créativ and creativ can be stored in the index, then it still won't work.

When I search for créativité, the analyzer will stem it and find créativ in the index. But when I search for creativite, it won't be stemmed since the stemmer looks for "ité" at the end. So since creativite is not in the index, no record will be returned.

Or do I miss something?

**Reply by ayende** [reply]
In your stemmer, you don't need to find valid words, or make it happen.
You can return multiple terms for the word. Including on query time.

So stem the word: "creativite" to "creativ", even if it is not valid.


Note that this is more about the way Lucene works than RavenDB directly. May find more resources in analyzers there.

**Reply by cadilhac** [reply]
Sorry, I really don't get it. If at query time, I create a kind of rule for this single one word "creativite", or words that finish in "ite" (and I'm sure I would create false positives in doing so), how will I do for the ton of other french words that have a similar issue? It seems simply impossible.

**Reply by ayende** [reply]
Basically, you create a set of options from the analyzer from the root word, and at query time from the term.
As long as there is any intersection, you find the data.

I would suggest asking about this issue in Lucene forums, since that is very specific to analyzer / scenario.

This one seems to be similar to your scenario:
https://stackoverflow.com/questions/45621411/lucene-combining-ascii-folding-and-stemming-for-french

The lemmas are here:
https://github.com/michmech/lemmatization-lists/

I'm afraid that this is not something that we are well suited to help here.

**Reply by cadilhac** [reply]
ok, thanks anyway.
